File: src/003/p003.rb
class Integer
  def prime?
    return false unless self > 1
    (2..Math.sqrt(self)).any? do |x|
      next if x == self
      return false if (self % x).zero?
    end
    true
  end

  def factors
    (1..Math.sqrt(self)).select { |x| (self % x).zero? }.each_with_object([]) do |number, list|
      list << number
      list << self / number unless self == number
    end
  end
end

def solve
  600_851_475_143.factors.select(&:prime?).max
end

puts solve

File: src/003/p003.exs
defmodule Problem003 do
  defp prime?(x) when x in [2, 3, 7, 11, 13, 17], do: true
  defp prime?(x) do
    Enum.all?((2..round(:math.sqrt(x))), fn(n) -> rem(x, n) != 0 end)
  end

  defp factors(n) do
    1..round(:math.sqrt(n))
    |> Enum.filter(fn(x) -> rem(n, x) == 0 end)
  end

  def solve do
    factors(600851475143)
    |> Enum.filter(&prime?/1)
    |> Enum.max
  end
end

IO.puts Problem003.solve
